src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -970,13 +1371,194 @@ fn slice_pat_covered_by_constructor<'tcx>(
+        if let ConstantValue(value) | ConstantRange(value, _, _) = ctor {
+            if let ty::TyChar | ty::TyInt(_) | ty::TyUint(_) = value.ty.sty {
+                return true;
+            }
+        }
+    }
+    false
+}
+
+/// For exhaustive integer matching, some constructors are grouped within other constructors
+/// (namely integer typed values are grouped within ranges). However, when specialising these
+/// constructors, we want to be specialising for the underlying constructors (the integers), not
+/// the groups (the ranges). Thus we need to split the groups up. Splitting them up naïvely would
+/// mean creating a separate constructor for every single value in the range, which is clearly
+/// impractical. However, observe that for some ranges of integers, the specialisation will be
+/// identical across all values in that range (i.e. there are equivalence classes of ranges of
+/// constructors based on their `is_useful_specialised` outcome). These classes are grouped by
+/// the patterns that apply to them (both in the matrix `P` and in the new row `p_{m + 1}`). We
+/// can split the range whenever the patterns that apply to that range (specifically: the patterns
+/// that *intersect* with that range) change.
+/// Our solution, therefore, is to split the range constructor into subranges at every single point
+/// the group of intersecting patterns changes, which we can compute by converting each pattern to
+/// a range and recording its endpoints, then creating subranges between each consecutive pair of
+/// endpoints.
+/// And voilà! We're testing precisely those ranges that we need to, without any exhaustive matching
+/// on actual integers. The nice thing about this is that the number of subranges is linear in the
+/// number of rows in the matrix (i.e. the number of cases in the `match` statement), so we don't
+/// need to be worried about matching over gargantuan ranges.
+fn split_grouped_constructors<'p, 'a: 'p, 'tcx: 'a>(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    ctors: Vec<Constructor<'tcx>>,
+    &Matrix(ref m): &Matrix<'p, 'tcx>,
+    p: &[&'p Pattern<'tcx>],
+    ty: Ty<'tcx>,
+) -> Vec<Constructor<'tcx>> {
+    let pat = &p[0];
+
+    let mut split_ctors = Vec::with_capacity(ctors.len());
+
+    for ctor in ctors.into_iter() {
+        match ctor {
+            // For now, only ranges may denote groups of "subconstructors", so we only need to
+            // special-case constant ranges.
+            ConstantRange(..) if should_treat_range_exhaustively(tcx, &ctor) => {
+                // We only care about finding all the subranges within the range of the intersection
+                // of the new pattern `p_({m + 1},1)` (here `pat`) and the constructor range.
+                // Anything else is irrelevant, because it is guaranteed to result in `NotUseful`,
+                // which is the default case anyway, and can be ignored.
+                let mut ctor_range = IntRange::from_ctor(tcx, &ctor).unwrap();
```

Comment:
This looks a bit suspicious. Shouldn't `ctors` already only include the constructors that are compatible with `p`?

Reply:
In which case can this be reached?

Reply:
They're going to be compatible in the sense that they're all intervals, but they might span much wider ranges than the pattern. I think we can potentially avoid this intersection if we want, but it eliminates some intervals at either side of the pattern.